src/bundle/Security/TwoFactor/Provider/TwoFactorProviderInitiator.php

```diff
@@ -21,46 +23,42 @@ public function __construct(
     ) {
     }
 
-    /**
-     * @return string[]
-     */
-    private function getActiveTwoFactorProviders(AuthenticationContextInterface $context): array
+    public function beginTwoFactorAuthentication(AuthenticationContextInterface $context): TwoFactorTokenInterface|null
     {
-        $activeTwoFactorProviders = [];
-
         // Iterate over two-factor providers and begin the two-factor authentication process.
+        $activeTwoFactorProviders = $statelessProviders = [];
         foreach ($this->providerRegistry->getAllProviders() as $providerName => $provider) {
             if (!$provider->beginAuthentication($context)) {
                 continue;
             }
 
             $activeTwoFactorProviders[] = $providerName;
+
+            if (!$provider->needsPreparation()) {
+                continue;
+            }
+
+            $statelessProviders[] = $providerName;
```

Comment:
I believe the logic is wrong in this place and that's why the integration test is failing. Actually it should be the other way round.

It's easy to get confused here, because the "return early" approach makes it harder to understand which values go into the array. Would rather implement it like this, to make the condition for the `$statelessProviders` array more clear:

```suggestion
            if (!$provider->needsPreparation()) {
                $statelessProviders[] = $providerName;
            }
```

Reply:
Check, my bad 👍